File: app/src/main/java/com/stardust/theme/preference/ThemeColorSwitchPreference.java
package com.stardust.theme.preference;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;
import android.preference.SwitchPreference;

import androidx.appcompat.widget.SwitchCompat;
import android.util.AttributeSet;
import android.view.View;
import android.widget.Checkable;
import android.widget.Switch;

import com.stardust.theme.ThemeColor;
import com.stardust.theme.ThemeColorHelper;
import com.stardust.theme.ThemeColorManager;
import com.stardust.theme.ThemeColorMutable;

/**
 * Created by Stardust on 2017/3/5.
 */
public class ThemeColorSwitchPreference extends SwitchPreference implements ThemeColorMutable {

    private View mCheckableView;
    private int mColor = Color.TRANSPARENT;

    public ThemeColorSwitchPreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init();
    }

    public ThemeColorSwitchPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public ThemeColorSwitchPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public ThemeColorSwitchPreference(Context context) {
        super(context);
        init();
    }

    private void init() {
        ThemeColorManager.add(this);
    }

    @Override
    protected void onBindView(View view) {
        super.onBindView(view);
        mCheckableView = view.findViewById(Resources.getSystem().getIdentifier("switch_widget", "id", "android"));

        if (mColor != Color.TRANSPARENT) {
            applyColor();
        }
    }

    @Override
    public void setThemeColor(ThemeColor color) {
        mColor = color.colorPrimary;
        if (mCheckableView != null) {
            applyColor();
        }
    }

    public void applyColor() {
        if (mCheckableView != null && mCheckableView instanceof Checkable) {
            if (mCheckableView instanceof final Switch switchView) {
                ThemeColorHelper.setColorPrimary(switchView, mColor);
            }

            if (mCheckableView instanceof final SwitchCompat switchView) {
                ThemeColorHelper.setColorPrimary(switchView, mColor);

            }
        }
    }
}

File: app/src/main/java/org/autojs/autojs/theme/preference/ThemeColorSwitchPreference.kt
package org.autojs.autojs.theme.preference

import android.content.Context
import android.content.res.Resources
import android.content.res.TypedArray
import android.graphics.Color
import android.util.AttributeSet
import android.view.View
import android.widget.Switch
import androidx.appcompat.widget.SwitchCompat
import androidx.preference.PreferenceViewHolder
import androidx.preference.SwitchPreference
import org.autojs.autojs.app.DialogUtils
import org.autojs.autojs.theme.ThemeColor
import org.autojs.autojs.theme.ThemeColorHelper
import org.autojs.autojs.theme.ThemeColorManager
import org.autojs.autojs.theme.ThemeColorMutable
import org.autojs.autojs6.R

/**
 * Created by Stardust on Mar 5, 2017.
 */
open class ThemeColorSwitchPreference : SwitchPreference, ThemeColorMutable, LongClickablePreferenceLike {

    private var mCheckableView: View? = null
    private var mColor = Color.TRANSPARENT

    override val prefTitle: CharSequence? = title
    override val prefContext: Context = context
    override var longClickPrompt: CharSequence? = null
    override var longClickPromptMore: CharSequence? = null

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(context, attrs, defStyleAttr, defStyleRes)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context, attrs, defStyleAttr, 0)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context, attrs, 0, 0)
    }

    constructor(context: Context) : super(context) {
        init(context, null, 0, 0)
    }

    init {
        ThemeColorManager.add(this)
    }

    private fun init(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) {
        obtainStyledAttrs(context, attrs, R.styleable.MaterialPreference, defStyleAttr, defStyleRes).let { a ->
            getAttrString(a, R.styleable.MaterialPreference_longClickPrompt)?.also { longClickPrompt = it }
            getAttrString(a, R.styleable.MaterialPreference_longClickPromptMore)?.also { longClickPromptMore = it }
            a.recycle()
        }
    }

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        mCheckableView = holder.findViewById(Resources.getSystem().getIdentifier("switch_widget", "id", "android"))
        if (mColor != Color.TRANSPARENT) {
            applyColor()
        }
        DialogUtils.applyLongClickability(this, holder)
    }

    override fun setThemeColor(color: ThemeColor) {
        mColor = color.colorPrimary
        applyColor()
    }

    private fun applyColor() {
        if (mCheckableView is Switch) {
            ThemeColorHelper.setColorPrimary(mCheckableView as Switch, mColor, true)
        }
        if (mCheckableView is SwitchCompat) {
            ThemeColorHelper.setColorPrimary(mCheckableView as SwitchCompat, mColor, true)
        }
    }

    protected fun obtainStyledAttrs(context: Context, set: AttributeSet?, styleableRes: IntArray, defStyleAttr: Int, defStyleRes: Int): TypedArray {
        return context.obtainStyledAttributes(set, styleableRes, defStyleAttr, defStyleRes)
    }

    protected fun getAttrString(a: TypedArray, index: Int): String? = a.getString(index)

}